ptychonn/_infer/__main__.py

```diff
@@ -0,0 +1,190 @@
+        Shrink the interpolation region by this number of pixels.
+    stitched_pixel_width : float [m]
+        The width of a pixel in the stitched image.
+    inference_pixel_width : float [m]
+        The width of a pixel in the inferred image patches.
+
+    Returns
+    -------
+    stitched : (COMBINED_WIDTH, COMBINED_HEIGHT) np.array
+        The stitched together image.
+    '''
+    pos_x = scan[..., 0]
+    pos_y = scan[..., 1]
+
+    # The global axes of the stitched image in meters
+    x = np.arange(pos_x.min(),
+                  pos_x.max() + inferences.shape[-1] * inference_pixel_width,
+                  step=stitched_pixel_width)
+    y = np.arange(pos_y.min(),
+                  pos_y.max() + inferences.shape[-1] * inference_pixel_width,
+                  step=stitched_pixel_width)
+
+    result = np.zeros((y.shape[0], x.shape[0]))
+    cnt = np.zeros_like(result)
+
+    # The local axes of the inference patches meters
+    xx = np.arange(inferences.shape[-1]) * inference_pixel_width
+
+    for i in tqdm.tqdm(range(len(inferences)), leave=False):
+        data_ = inferences[i]
+        xxx = xx + pos_x[i]
+        yyy = xx + pos_y[i]
+
+        if pix > 0:
+            xxx = xxx[pix:-pix]
+            yyy = yyy[pix:-pix]
+            data_ = data_[pix:-pix, pix:-pix]
+        find_pha = scipy.interpolate.interp2d(
+            xxx,
+            yyy,
+            data_,
+            kind='linear',
+            fill_value=0,
+        )
+        tmp = find_pha(x, y)
+        cnt += tmp != 0
+        result += tmp
+
+    return result / np.maximum(cnt, 1)
+
+
+@click.command(name='infer')
+@click.argument(
+    'data_path',
+    type=click.Path(
+        exists=True,
+        path_type=pathlib.Path,
+    ),
+)
+@click.argument(
+    'scan_path',
+    type=click.Path(
+        exists=True,
+        path_type=pathlib.Path,
+    ),
+)
+@click.argument(
+    'out_dir',
+    type=click.Path(
+        exists=True,
+        path_type=pathlib.Path,
+    ),
+)
+def infer_cli(
+    data_path: pathlib.Path,
+    scan_path: pathlib.Path,
+    out_dir: pathlib.Path,
+):
+    '''Infer a reconstructed image from diffraction patterns at DATA_PATH and
+    scan positions at SCAN_PATH. Save inferred patches and stitched image at
+    OUT_DIR.
+    '''
+
+    inferences_out_file = out_dir / 'inferences_506.npz'
+    click.echo(f'Does data path exist? {data_path.exists()}')
+
+    with h5py.File(data_path) as f:
+        data = f['entry/data/data'][()]
+
+    inferences = infer(
+        data=data,
+        inferences_out_file=inferences_out_file,
+    )
+
+    ## parameters required for stitching individual inferences
+    spiral_step = 0.05
+    step = spiral_step * -1e-6
+    spiral_traj = np.load(scan_path)
+    scan = np.stack((spiral_traj['x'], spiral_traj['y']), axis=-1) * step
+    stitched = stitch_from_inference(
+        inferences,
+        scan,
+        stitched_pixel_width=10e-9,
+        inference_pixel_width=10e-9,
+    )
```

Comment:
Start here to understand the workflow. At a high level, the workflow steps are:

1. Load diffraction patterns
2. Give the diffraction patterns to the trained network for inference
3. Stitch the network output together into a full reconstructed image